### ISSUE TITLE
Follower can't delete because Segment is not open

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#3885]
**Describe the bug**

I had multiple benchmarks now where I saw that an follower is not able to delete/compact the log because it is not able to create an reader to seek to a position. The cause seems that the related segment is not open?

```
I 2020-02-20T04:36:35.764188706Z 2020-02-20 04:36:35.763 [] [raft-server-1-raft-partition-partition-3] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(org.camunda.feel.spi.JavaValueMapper@2b208a60)), function-provider: org.camunda.feel.interpreter.FunctionProvider$EmptyFunctionProvider$@35052193]
 
I 2020-02-20T04:36:35.767904147Z 2020-02-20 04:36:35.767 [] [raft-server-1-raft-partition-partition-3] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - The last processed position for snapshot /usr/local/zeebe/data/raft-partition/partitions/3/pending/58622542-5-1582173300964 is 140569984646512
 
I 2020-02-20T04:36:35.775736343Z 2020-02-20 04:36:35.775 [] [raft-server-1-raft-partition-partition-3] INFO  io.zeebe.broker.clustering.atomix.storage.snapshot.AtomixSnapshotStorage - Max snapshot count reached (2), purging snapshots older than DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/58622542-5-1582173300964-140569984646512, metadata=DbSnapshotMetadata{index=58622542, term=5, timestamp=2020-02-20 04:35:00,964, position=140569984646512}}
 
I 2020-02-20T04:36:35.775904072Z 2020-02-20 04:36:35.775 [] [raft-server-1-raft-partition-partition-3] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Deleting snapshot DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/57582033-5-1582172415688-137559214458800, metadata=DbSnapshotMetadata{index=57582033, term=5, timestamp=2020-02-20 04:20:15,688, position=137559214458800}}
 
I 2020-02-20T04:36:35.917118395Z 2020-02-20 04:36:35.916 [] [raft-server-1-raft-partition-partition-3] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.AtomixSnapshotStorage - Snapshot DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/57582033-5-1582172415688-137559214458800, metadata=DbSnapshotMetadata{index=57582033, term=5, timestamp=2020-02-20 04:20:15,688, position=137559214458800}} removed from store io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore@32f489f8
 
I 2020-02-20T04:36:35.917176460Z 2020-02-20 04:36:35.917 [] [raft-server-1-raft-partition-partition-3] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Search for orphaned snapshots below oldest valid snapshot DbSnapshotMetadata{index=58622542, term=5, timestamp=2020-02-20 04:35:00,964, position=140569984646512}
 
I 2020-02-20T04:36:35.917482867Z 2020-02-20 04:36:35.917 [] [raft-server-1-raft-partition-partition-3] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Committed new snapshot DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/58622542-5-1582173300964-140569984646512, metadata=DbSnapshotMetadata{index=58622542, term=5, timestamp=2020-02-20 04:35:00,964, position=140569984646512}}
 
I 2020-02-20T04:36:36.011498958Z 2020-02-20 04:36:36.011 [Broker-1-DeletionService-3] [Broker-1-zb-actors-3] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(org.camunda.feel.spi.JavaValueMapper@77f8b997)), function-provider: org.camunda.feel.interpreter.FunctionProvider$EmptyFunctionProvider$@35052193]
 
I 2020-02-20T04:36:36.014528047Z 2020-02-20 04:36:36.014 [Broker-1-DeletionService-3] [Broker-1-zb-actors-3] DEBUG io.zeebe.broker.system - The last processed position for snapshot /usr/local/zeebe/data/raft-partition/partitions/3/snapshots/58622542-5-1582173300964-140569984646512 is 140569984646512
 
I 2020-02-20T04:36:36.014813021Z 2020-02-20 04:36:36.014 [Broker-1-DeletionService-3] [Broker-1-zb-actors-3] DEBUG io.zeebe.broker.system - The lowest exported position for snapshot /usr/local/zeebe/data/raft-partition/partitions/3/snapshots/58622542-5-1582173300964-140569984646512 is 140565697448456
 
I 2020-02-20T04:36:36.019152479Z 2020-02-20 04:36:36.018 [Broker-0-LogStream-3] [Broker-1-zb-actors-3] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
 
I 2020-02-20T04:36:36.019177802Z java.lang.IllegalStateException: Segment not open
	at com.google.common.base.Preconditions.checkState(Preconditions.java:508) ~[guava-28.2-jre.jar:?]
	at io.atomix.storage.journal.JournalSegment.checkOpen(JournalSegment.java:237) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.atomix.storage.journal.JournalSegment.createReader(JournalSegment.java:213) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.atomix.storage.journal.SegmentedJournalReader.replaceCurrentSegment(SegmentedJournalReader.java:204) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.atomix.storage.journal.SegmentedJournalReader.reset(SegmentedJournalReader.java:128) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.atomix.storage.journal.SegmentedJournalReader.reset(SegmentedJournalReader.java:136) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.atomix.storage.journal.DelegatingJournalReader.reset(DelegatingJournalReader.java:77) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.zeebe.logstreams.storage.atomix.AtomixLogStorageReader.findEntry(AtomixLogStorageReader.java:166) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.logstreams.storage.atomix.AtomixLogStorageReader.lookUpApproximateAddress(AtomixLogStorageReader.java:123) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.logstreams.impl.log.LogStreamReaderImpl.seek(LogStreamReaderImpl.java:70) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.logstreams.impl.log.LogStreamImpl.lambda$delete$4(LogStreamImpl.java:167) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorControl.lambda$call$0(ActorControl.java:150) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:62) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
 
I 2020-02-20T04:37:14.845481181Z 2020-02-20 04:37:14.844 [] [raft-server-1-raft-partition-partition-1] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(org.camunda.feel.spi.JavaValueMapper@3b99f986)), function-provider: org.camunda.feel.interpreter.FunctionProvider$EmptyFunctionProvider$@35052193]
 
I 2020-02-20T04:37:14.847224304Z 2020-02-20 04:37:14.846 [] [raft-server-1-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - The last processed position for snapshot /usr/local/zeebe/data/raft-partition/partitions/1/pending/51438185-6-1582173336362 is 121766618320960
 
I 2020-02-20T04:37:14.857835324Z 2020-02-20 04:37:14.857 [] [raft-server-1-raft-partition-partition-1] INFO  io.zeebe.broker.clustering.atomix.storage.snapshot.AtomixSnapshotStorage - Max snapshot count reached (2), purging snapshots older than DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/51438185-6-1582173336362-121766618320960, metadata=DbSnapshotMetadata{index=51438185, term=6, timestamp=2020-02-20 04:35:36,362, position=121766618320960}}
 
I 2020-02-20T04:37:14.858216046Z 2020-02-20 04:37:14.858 [] [raft-server-1-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Deleting snapshot DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/50365151-6-1582172457497-118652772929608, metadata=DbSnapshotMetadata{index=50365151, term=6, timestamp=2020-02-20 04:20:57,497, position=118652772929608}}
 
I 2020-02-20T04:37:14.979955507Z 2020-02-20 04:37:14.979 [] [raft-server-1-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.AtomixSnapshotStorage - Snapshot DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/50365151-6-1582172457497-118652772929608, metadata=DbSnapshotMetadata{index=50365151, term=6, timestamp=2020-02-20 04:20:57,497, position=118652772929608}} removed from store io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore@37cf9ba3
 
I 2020-02-20T04:37:14.979991610Z 2020-02-20 04:37:14.979 [] [raft-server-1-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Search for orphaned snapshots below oldest valid snapshot DbSnapshotMetadata{index=51438185, term=6, timestamp=2020-02-20 04:35:36,362, position=121766618320960}
 
I 2020-02-20T04:37:14.980286887Z 2020-02-20 04:37:14.980 [] [raft-server-1-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Committed new snapshot DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/51438185-6-1582173336362-121766618320960, metadata=DbSnapshotMetadata{index=51438185, term=6, timestamp=2020-02-20 04:35:36,362, position=121766618320960}}
 
I 2020-02-20T04:37:15.092601563Z 2020-02-20 04:37:15.092 [Broker-1-DeletionService-1] [Broker-1-zb-actors-0] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(org.camunda.feel.spi.JavaValueMapper@f1f8262)), function-provider: org.camunda.feel.interpreter.FunctionProvider$EmptyFunctionProvider$@35052193]
 
I 2020-02-20T04:37:15.095733596Z 2020-02-20 04:37:15.095 [Broker-1-DeletionService-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.system - The last processed position for snapshot /usr/local/zeebe/data/raft-partition/partitions/1/snapshots/51438185-6-1582173336362-121766618320960 is 121766618320960
 
I 2020-02-20T04:37:15.096260153Z 2020-02-20 04:37:15.096 [Broker-1-DeletionService-1] [Broker-1-zb-actors-0] DEBUG io.zeebe.broker.system - The lowest exported position for snapshot /usr/local/zeebe/data/raft-partition/partitions/1/snapshots/51438185-6-1582173336362-121766618320960 is 121762326174248
 
I 2020-02-20T04:37:15.101574770Z 2020-02-20 04:37:15.101 [Broker-0-LogStream-1] [Broker-1-zb-actors-0] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
 
I 2020-02-20T04:37:15.101602705Z java.lang.IllegalStateException: Segment not open
	at com.google.common.base.Preconditions.checkState(Preconditions.java:508) ~[guava-28.2-jre.jar:?]
	at io.atomix.storage.journal.JournalSegment.checkOpen(JournalSegment.java:237) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.atomix.storage.journal.JournalSegment.createReader(JournalSegment.java:213) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.atomix.storage.journal.SegmentedJournalReader.replaceCurrentSegment(SegmentedJournalReader.java:204) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.atomix.storage.journal.SegmentedJournalReader.reset(SegmentedJournalReader.java:128) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.atomix.storage.journal.SegmentedJournalReader.reset(SegmentedJournalReader.java:136) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.atomix.storage.journal.DelegatingJournalReader.reset(DelegatingJournalReader.java:77) ~[atomix-storage-3.2.0-20200214.153035-125.jar:?]
	at io.zeebe.logstreams.storage.atomix.AtomixLogStorageReader.findEntry(AtomixLogStorageReader.java:166) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.logstreams.storage.atomix.AtomixLogStorageReader.lookUpApproximateAddress(AtomixLogStorageReader.java:123) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.logstreams.impl.log.LogStreamReaderImpl.seek(LogStreamReaderImpl.java:70) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.logstreams.impl.log.LogStreamImpl.lambda$delete$4(LogStreamImpl.java:167) ~[zeebe-logstreams-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorControl.lambda$call$0(ActorControl.java:150) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:62) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
 
I 2020-02-20T04:37:30.945458268Z 2020-02-20 04:37:30.945 [GatewayLongPollingJobHandler] [Broker-1-zb-actors-0] DEBUG io.zeebe.gateway - Remove blocking request BrokerActivateJobsRequest{requestDto={"type":"benchmark-task","worker":"benchmark-worker","timeout":1800,"maxJobsToActivate":84,"jobKeys":[],"jobs":[],"variables":[],"truncated":false}} for job type benchmark-task after timeout of PT1M2S
 
```

Happend on https://github.com/zeebe-io/zeebe/issues/3760#issuecomment-589020702

**Comment by npepinpe** [comment]
How does the memory look? I've run into this issue if you get (silently ignored :see_no_evil:) OOM errors, as the code to move to the next/previous segment first closes the current segment reader before allocating the new segment reader; if an OOM is thrown, it will then keep chugging along but with an invalid state where it's now impossible to read anything.

**Comment by Zelldon** [comment]
Related metrics for that benchmark https://github.com/zeebe-io/zeebe/issues/3760#issuecomment-589020702

**Comment by npepinpe** [comment]
Ah, so probably unrelated then =/

**Comment by npepinpe** [comment]
There was also a case where it's possible for the follower to receive a snapshot (because it is slow), and attempt to delete before it even receives the next events, resulting in a failed compaction (e.g. the event to delete from is not found because it hasn't been replicated yet). This is a minor problem, but one that's directly due to Atomix assuming that a snapshot represents the projection of the log up until the given index, and does not **need** that index to be able to recover (as it has no concept of reprocessing).

**Comment by npepinpe** [comment]
Closed by #3962 - we don't need to find the exact entry anymore, so this should not happen again.
